Server UI improvements

mounta11n opened the pull request:
- Make UI look nicer and cleaner overall with CSS.
- CSS outsourced as a separate file
- Added a dropdown menu with prompt style templates
- Added a dropdown menu with system prompts
- Prompt Styles and System Prompts are separate files, so editing is very easy.
- User friendly TUI: Created a script that uses "dialog" to compose the command for the server.
- Dialog Script offers the possibility to save and load configs

---

In planning or already started:

- WIP Multilingual: You will be able to select the language from a dropdown menu. So far language files only for English and German. (concerns UI elements and system prompts).
- Dark Mode
- Templates for the values of the UI options (samplers etc.), e.g. deterministic template, creative template, balanced template etc...

![ppCamall-ui01](https://github.com/ggerganov/llama.cpp/assets/123123600/b50ae0a6-b387-4f7f-afaf-361b07d178df)
![ppCamall-ui02](https://github.com/ggerganov/llama.cpp/assets/123123600/203433b8-6438-4578-b153-1e631e0990e1)
![ppCamall-tui01](https://github.com/ggerganov/llama.cpp/assets/123123600/b91da5b5-6282-4778-bf66-0c62b3f1193f)

Commit:
ic

Commit:
add start script

mounta11n committed:
Update index.html

add atlas

Commit:
add favicon

mounta11n committed:
Merge pull request #1 from mounta11n/server-ui-improvements

Server UI improvements

mounta11n committed:
Update README.md

Commit:
add ui and tui pics

Commit:
add ui and tui pics

mounta11n committed:
Update README.md

add pictures of the ui

mounta11n committed:
Update README.md

add description

ggerganov commented:
Cool stuff! The `dialog` TUI seems very handy - will learn how it works.

Do you plan to keep updating this PR or you prefer this to get merged?

Commit:
fix multi-modal-selection

ggerganov commented:
> I see that there are some formatting issues. Let me know if you want me to correct them first. Otherwise I would be happy if you merge it.

Yup, try to fix these. I use `editorconfig-checker` locally like this:

```bash
brew install editorconfig editorconfig-checker
editorconfig-checker
```

ggerganov commented on examples/server/public/index.html:
```diff
@@ -419,17 +230,15 @@
           slot_id = data.slot_id;
           if (selected_image && !data.multimodal) {
-            alert("The server was not compiled for multimodal or the model projector can't be loaded.");
+            alert("Der Server wurde nicht für multimodal zusammengestellt oder der Modellprojektor kann nicht geladen werden.");
```
Should we keep these errors in English?

Commit:
fix wrong formattings

ggerganov commented:
Will let the PR stay open for a ~day so we gather some additional feedback. Would be great if people run some tests in the meantime to make sure it works OK

Commit:
fix wrong translation

Commit:
add mmproj function

mounta11n committed:
Merge branch 'ggerganov:master' into master

mounta11n committed:
Merge pull request #2 from mounta11n/server-ui-improvements

Server UI improvements

Commit:
add min-p

mounta11n committed:
Merge pull request #3 from mounta11n/server-ui-improvements

add min-p

mounta11n committed:
Update README.md

Commit:
add min-p image

mounta11n committed:
Update README.md

mounta11n committed:
Merge branch 'ggerganov:master' into master

Commit:
add start scripts to root path

Commit:
update

Commit:
error handling for missing dialog

Commit:
fix start-server.sh

Commit:
update start server scripts

mounta11n committed:
Update promptFormats.js

mounta11n committed:
Update promptFormats.js

Commit:
improve error handling

mounta11n committed:
Merge branch 'ggerganov:master' into master

mounta11n committed:
fix typo

mounta11n committed:
Merge branch 'master' into server-ui-improvements

ggerganov commented:
I just tested this and on my browser it defaults to dark mode which does not look good:

![image](https://github.com/ggerganov/llama.cpp/assets/1991296/2a836333-826b-4727-a4c6-f5c5f10ad223)

Left is new, right is old

- Some of the text cannot be read
- Sliders are hard to see

Would it be possible to add UI control to select the style and be able to switch back to the old CSS?

mounta11n commented:
That's because the colors for dark mode in the style.css file should indicate that there is this work in progress. Think of them as a placeholder for now.

The colors were absolutely not actively chosen. Also I think this specific green is maybe not the best choice for dark mode, what do you think? I was initially thinking to make it look like your llama.cpp banner (I mean in terms of colors; the black/dark gray and orange)

![230134379-7181e485-c521-4d23-a0d6-f7b3b61ba524](https://github.com/ggerganov/llama.cpp/assets/123123600/f0fa3541-9633-4ba8-a13d-734eb9570e26)


The green color of your whisper banner or something similiar would also fit well to a dark mode i find, but i really like the orange very much. And it will give the whole thing a recognizable value.


And one more thing I want to get rid of: This is my first 'serious' pull request, so I'm not very practiced or used to the concept, so I wasn't sure how to handle it overall. Specifically, I still don't know if it's common practice that you should submit a pull request as a finished packet, so to speak, and then say goodbye if you want to. Or whether it is traditionally expected that the person who makes a pull request ideally also takes care of it afterwards or at least makes sure that the code is well maintained. I initially assumed the latter, but I can also imagine the former. So long story short :D Please tell me concretely how it should ideally be. Also if there are other things that I haven't noticed. 


Regardless of that, I'll continue working on the dark mode right away and see that I finish it as soon as I can. It shouldn't take that long. So to your question: Yes, it's definitely possible and I can do it right away.

Just please give me a short feedback regarding the accent color. Do you agree with the orange or do you have certain criteria that should be met?

ggerganov commented:
> Specifically, I still don't know if it's common practice that you should submit a pull request as a finished packet, so to speak, and then say goodbye if you want to. Or whether it is traditionally expected that the person who makes a pull request ideally also takes care of it afterwards or at least makes sure that the code is well maintained.

There are no "strings attached" - if you don't have time or will to contribute in the future, that's completely fine. We don't expect any long or short term commitment and any contribution, big or small, is very welcome.

I don't have a strong preference of the style - as long as it renders OK. It does make sense to reuse the orange - yes.

> Regardless of that, I'll continue working on the dark mode right away and see that I finish it as soon as I can.

No rush, have fun!

Huge commented:
I have posted some feedback/experience under https://www.reddit.com/r/LocalLLaMA/comments/18534f1/i_have_given_llamacpp_server_ui_a_facelift/

- I am not able to select and copy any generated text while still generating. Seems like a UX bug where the selection disappears with each token streamed in.

- possibly wrong suggestion, but I [got used](https://github.com/abetlen/llama-cpp-python) to have /docs endpoint with description of the endpoints available.
  It could point to/render https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md#api-endpoints at first, anyway seems helpful to have it served.
  
-   the prob. values seem [off, not normalized](https://preview.redd.it/rmmx9kclh33c1.png?width=183&format=png&auto=webp&s=03a94c6b8d21dee12ab9822f4206752296e08172), sometimes.

Azeirah commented:
> I am not able to select and copy any generated text while still generating. Seems like a UX bug where the selection disappears with each token streamed in.

I noticed that too, it's not a new bug introduced by this PR though. I'm not sure how feasible it is to implement copy paste functionality on text that gets replaced this frequently.

I'll take a look at it though, there's always the possibility of caching the selection and just reapplying it after re-render, but I'm prettyyy sure that's going to have some nasty side-effects.

Azeirah commented:
> > I am not able to select and copy any generated text while still generating. Seems like a UX bug where the selection disappears with each token streamed in.
> 
> I noticed that too, it's not a new bug introduced by this PR though. I'm not sure how feasible it is to implement copy paste functionality on text that gets replaced this frequently.
> 
> I'll take a look at it though, there's always the possibility of caching the selection and just reapplying it after re-render, but I'm prettyyy sure that's going to have some nasty side-effects.

Fixing it in vanilla JS turned out to be pretty simple, instead of mutating the innerText/textContent of an element, it's easier to append a Text node to the container element. This way the selection is retained because nothing is rerendered, only appended.

IE: https://codepen.io/laauurraaa/pen/MWLPJed?editors=1010

How this works in the current server UI I'm not too certain, I haven't figured out what framework it's even using. It's using a minified version of a very lightweight React-like? Possibly Preact or SolidJS? I'm not familiar with those.

tobi commented:
I like the ux updates!

The chat template system is a bit conflicting with the work/thought going on at https://github.com/ggerganov/llama.cpp/issues/4216 - basically, what we should do is to make the client/html dumber and have it use the openai endpoint to get chat automatically. We want the endpoint to apply the chat template that the model requires automatically.

mounta11n commented:
> There are no "strings attached" - if you don't have time or will to contribute in the future, that's completely fine. We don't expect any long or short term commitment and any contribution, big or small, is very welcome.
> 
> I don't have a strong preference of the style - as long as it renders OK. It does make sense to reuse the orange - yes.
> 
> > Regardless of that, I'll continue working on the dark mode right away and see that I finish it as soon as I can.
> 
> No rush, have fun!

Okay thanks


> I have posted some feedback/experience under https://www.reddit.com/r/LocalLLaMA/comments/18534f1/i_have_given_llamacpp_server_ui_a_facelift/
> 
>     * I am not able to select and copy any generated text while still generating. Seems like a UX bug where the selection disappears with each token streamed in.
> 
>     * possibly wrong suggestion, but I [got used](https://github.com/abetlen/llama-cpp-python) to have /docs endpoint with description of the endpoints available.
>       It could point to/render https://github.com/ggerganov/llama.cpp/blob/master/examples/server/README.md#api-endpoints at first, anyway seems helpful to have it served.
> 
>     * the prob. values seem [off, not normalized](https://preview.redd.it/rmmx9kclh33c1.png?width=183&format=png&auto=webp&s=03a94c6b8d21dee12ab9822f4206752296e08172), sometimes.

Thanks for the feedback. I am not sure if I have understood you correctly. Do you mean that after you press submit/send and while the llm is generating, you are not able to select text already generated?

If so, I couldn't reproduce this error. If not, please elaborate how I could reproduce it. I will anyway investigate and see if this bug is caused by my code and if I could fix it.
I am just debugging some other UI issues, but I am almost finished with it and will start investigating right away after that.


@all fyi:
While I was working on to implement a dark mode, I decided to change my goal and instead of 
creating a dark mode, I am creating various themes, so one could have a more personalized UI, and it is now very easy to create own color-themes.


Here is a first impression (additionally to the OpenAI-Playground-like UI):

<br>
<br>

![ui_polar-1_20231201034003800](https://github.com/ggerganov/llama.cpp/assets/123123600/ef2c8827-3db9-4294-9a3d-96e79652867d)

<br>
<br>

![ui_snow-1_20231201033808342](https://github.com/ggerganov/llama.cpp/assets/123123600/6bf6e236-34bd-47a9-9052-3f0a73e81877)

<br>
<br>

In case someone is interested. The following output was generated by starling 7b, which indeed surprised me on how good this model is. Its understanding and producing of german language is impressive considering its size and english focused dataset.
however its coding capabilities are poor, for example it struggles with easy css or html tasks.

![iScreen Shoter - Safari - 231201032226](https://github.com/ggerganov/llama.cpp/assets/123123600/a255443c-2b71-4f8c-a71d-b47f08692f07)



> Fixing it in vanilla JS turned out to be pretty simple, instead of mutating the innerText/textContent of an element, it's easier to append a Text node to the container element. This way the selection is retained because nothing is rerendered, only appended.
> 
> IE: https://codepen.io/laauurraaa/pen/MWLPJed?editors=1010

Okay, that sounds good, but it is interesting that I could not reproduce this behaviour. I'll try your code out.

> I like the ux updates!
> 
> The chat template system is a bit conflicting with the work/thought going on at #4216 - basically, what we should do is to make the client/html dumber and have it use the openai endpoint to get chat automatically. We want the endpoint to apply the chat template that the model requires automatically.

Oh thanks for the hint. Okay I will read what is going on at #4216 and address what you mentioned. by applying the templates automatically, do you mean to get the information from the .gguf itself? I was thinking about something like that, but found that in some models the information was missing. But anyway, I can follow your works and orient myself accordingly.

Tostino commented:
@mounta11n It means that the UI wouldn't have to deal with the template at all, you just need to hit the `chat/completions` endpoint with a list of messages with a role:content pairs, and the server will handle templating it out and returning just the model's response.

For most models, they seem to stick to the `system`, `assistant`, and `user` roles...but that is far from guaranteed.
For example, my model has a few extra roles that hydrate some extra variables in the template if they are passed in (`meta-current_date`, `meta-task_name`), and then I also have an additional role for any content that the user is asking about (`user_context`) that can be interleaved with the `user` and `assistant` messages.

e.g.
```
curl http://0.0.0.0:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer " \
  -d '{
    "model": "Tostino/Inkbot-13B-8k-0.2",
    "stop": ["\n<#user#>", "\n<#bot#>"],
    "stream": false,
    "add_generation_prompt": true,
    "echo": false,
    "n": 1,
    "messages": [
	{"role": "meta-current_date", "content": "2023-10-20"},
	{"role": "meta-task_name", "content": "general"},
	{"role": "system", "content": "You are a helpful assistant."},
	{"role": "user", "content": "What color is the car?"},
	{"role": "user_context", "content": "The car is blue."}
	]
  }'
  ```


I'm not saying "implement support for my use case" here, I just wanted to bring it up so you are aware of what is possible.

You can see how the endpoint for [vLLM](https://github.com/vllm-project/vllm/pull/1756) that I implemented works, fingers crossed when llama.cpp implements it, they try and stay compatible with the choices that were made.

mounta11n committed:
Update README.md

mounta11n committed:
Merge branch 'ggerganov:master' into server-ui-improvements

mounta11n committed:
add dark and light color themes

mounta11n commented:
Here is some little update. the dark mode/light mode work is finished and here are the various themes:

## Light Themes

<br>
<figure >
  <img src="https://github.com/ggerganov/llama.cpp/assets/123123600/04fecb8b-29c4-4754-8e8d-86c122693176" width="800" />
  <figcaption>Default and Boring, but could maybe a good starting point for customizing own themes</figcaption>
</figure>
<br>

<br>
<figure >
  <img src="https://github.com/ggerganov/llama.cpp/assets/123123600/127ea49c-aaa7-465a-9776-d827f9775f49" width="800" />
  <figcaption>The already shown OAI-Playground-like theme</figcaption>
</figure>
<br>

<br>
<figure >
  <img src="https://github.com/ggerganov/llama.cpp/assets/123123600/3786b6e0-b429-4d91-96e6-0ad052a593b1" width="800" />
  <figcaption>Looks professional</figcaption>
</figure>
<br>
<br>

## Dark Themes

<br>
<figure >
  <img src="https://github.com/ggerganov/llama.cpp/assets/123123600/36881238-f25b-443b-b7bb-05574b47632e" width="800" />
  <figcaption>Mango Tango - inspired bei llama.cpp logo/banner</figcaption>
</figure>
<br>

<br>
<figure >
  <img src="https://github.com/ggerganov/llama.cpp/assets/123123600/b226e071-f071-49a2-a5db-d07ecaab5759" width="800" />
  <figcaption>Nord Theme</figcaption>
</figure>
<br>

<br>
<figure >
  <img src="https://github.com/ggerganov/llama.cpp/assets/123123600/28b93715-5862-4e20-a14e-21a654cec8eb" width="800" />
  <figcaption>favorite</figcaption>
</figure>
<br>


---

As already mentioned, the work on the UI elements and themes has been completed. Now it's on to the next work: figuring out how to fix the misbehaviors mentioned by other users (plus I've discovered a few more bugs in the meantime). This mainly occurs when toggling to completion and then back to chat. I would appreciate any support, especially if someone who is very familiar with Javascript and has already done a " eye diagnosis" could give me some advice.

Azeirah commented:
> figuring out how to fix the misbehaviors mentioned by other users (plus I've discovered a few more bugs in the meantime). This mainly occurs when toggling to completion and then back to chat. I would appreciate any support, especially if someone who is very familiar with Javascript and has already done a " eye diagnosis" could give me some advice.

I can pick up the JS issues. I think it's best to merge this as-is, you haven't introduced any JS bugs as far as I know?

We can open a new issue and list the bugs you're aware of. I'm a web dev and I can look at the JS parts and maybe others can look at it too.

That way your styles become available for everyone to use, and you don't have to worry about the JS bugs and they'll be fixed soon regardless!

mounta11n commented:
> I can pick up the JS issues. I think it's best to merge this as-is, you haven't introduced any JS bugs as far as I know?
> 
> We can open a new issue and list the bugs you're aware of. I'm a web dev and I can look at the JS parts and maybe others can look at it too.
> 
> That way your styles become available for everyone to use, and you don't have to worry about the JS bugs and they'll be fixed soon regardless!

Oh that would be very helpful indeed! I would be very pleased about that. Do I have to take any specific action now?

ggerganov commented:
- When I toggle between "Chat" and "Complete", the style control stops working
- I was hoping that the "Default" style would remain as the one on `master`

Regarding templates - probably this PR would need to be split into UI changes and template changes. The goal is to have the server apply the correct chat template automatically, so the template / prompt functionality on the client would eventually get deprecated.

mounta11n commented:
> > > I can pick up the JS issues. I think it's best to merge this as-is, you haven't introduced any JS bugs as far as I know?
> > > We can open a new issue and list the bugs you're aware of. I'm a web dev and I can look at the JS parts and maybe others can look at it too.
> > > That way your styles become available for everyone to use, and you don't have to worry about the JS bugs and they'll be fixed soon regardless!
> > 
> > 
> > Oh that would be very helpful indeed! I would be very pleased about that. Do I have to take any specific action now?
> 
> I suppose wait for @ggerganov for feedback/merging. Other than that, it's helpful if you could open an issue with a list of the bugs present in the UI with steps on how to reproduce the bugs

Just to be sure if I understand you right and not gonna doing something dumb: Do you mean open an issue here in gerganov's repo, even this pr is not merged yet? I am asking because for now I/we do not know if the at least some of the bugs are the result of my code. So I would open an issue with the bugs I know and as a _sidenote_ refer to this pr, right?

PS: one bug that I noticed is described by ggerganov, seens to occur there too.


> * When I toggle between "Chat" and "Complete", the style control stops working
> 
> * I was hoping that the "Default" style would remain as the one on `master`
> 
> 
> Regarding templates - probably this PR would need to be split into UI changes and template changes. The goal is to have the server apply the correct chat template automatically, so the template / prompt functionality on the client would eventually get deprecated.

I am sorry, I didn't understand your second point. What do you mean by "Default" style? Do you mean that one I introduced in my last post and which I described as 'boring'? Or do you mean the original style, the one before I made this pr at all? Or the oai-playground-like one, the one I have on my fork/master?


Yes, I have heard that there are discussions about integrating the prompt templates into the program logic and standardizing the prompt interface. We can do it the way you suggest if you think that's better.

My thought on this is that it would make sense to keep it here in this pull request after all, because this is a long awaited feature (e.g. the resonance on this pr in reddit LocalLlama was very positive and people are waiting for the merge there, in large part because the prompt style templates are expected).

The other point is that a second pr will probably not be necessary if a different and more elegant solution is already being worked on anyway. So my suggestion would be that the prompt style templates remain temporary to bridge the gap until the other project is ready. There could be a note on the server README.md (and perhaps also in the UI itself) that the current implementation of the prompt styles will very probably soon be obsolete.

Let me know how I should proceed : )

Aside from that, just want to double check: Regarding the system prompts, do we agree that they should stay and that they fulfill a useful function in the frontend?

ps: in the meantime i have continued to work on the UI and replaced the dropdown menu for the color styles with a nicer css menu and also changed the layout slightly. e.g. the color styles menu has moved up two 'sections' because it has a very superordinate role. Ironically, as a result, the bug where the styles could no longer be changed after toggling between chat and completion is no longer reproducible.

![css-menu-colors](https://github.com/ggerganov/llama.cpp/assets/123123600/8060459f-df3f-4bc5-9cc0-99bd406bd871)

ggerganov commented:
> Just to be sure if I understand you right and not gonna doing something dumb: Do you mean open an issue here in gerganov's repo, even this pr is not merged yet? I am asking because for now I/we do not know if the at least some of the bugs are the result of my code. So I would open an issue with the bugs I know and as a sidenote refer to this pr, right?

The one that is on the `master` branch, before this PR.

![image](https://github.com/ggerganov/llama.cpp/assets/1991296/7fcbbac0-5a1e-4cb3-9f10-6e5b38c94f90)

> Aside from that, just want to double check: Regarding the system prompts, do we agree that they should stay and that they fulfill a useful function in the frontend?

I think so. From what I gather, the system prompt can be used with some of the chat templates.

> Yes, I have heard that there are discussions about integrating the prompt templates into the program logic and standardizing the prompt interface. We can do it the way you suggest if you think that's better.

> My thought on this is that it would make sense to keep it here in this pull request after all, because this is a long awaited feature (e.g. the resonance on this pr in reddit LocalLlama was very positive and people are waiting for the merge there, in large part because the prompt style templates are expected).

I don't have a strong opinion on this. If the frontend logic for supporting the templates is lightweight enough and can be easily deprecated in the future, then I guess it is fine.

mounta11n committed:
Merge remote-tracking branch

mounta11n committed:
debugging

mounta11n committed:
fix mistral prompt

mounta11n committed:
sync gitignore

mounta11n committed:
clean up & split PRs

mounta11n commented:
@ggerganov sorry for the delayed response. In the meantime I've decided to split this PR into two. One for the server improvements and another with the dialog-start-server script.

The current state of things is that I have left the old UI as the default one and as index.html allmost untouched, beside that I've added a linking button to the new UI as you can see in the image. The same is with the new UI, which is now index-new.html and it has a link to the old UI.

Please tell me if there is something else to do in the front end. I will try to hurry up, cause I pretty much hope to get the new ui merged before christmas : D

![Safari-231221022105](https://github.com/ggerganov/llama.cpp/assets/123123600/cbea1d86-a124-47a5-a6f8-49491284ac61)
![Safari-231221022152](https://github.com/ggerganov/llama.cpp/assets/123123600/0c83ed35-bfa9-4f91-95d6-39f07707801a)

mounta11n commented:
And oh, I don't know what kind of mess I made with the files shown as changed. Actually it should **only** affect files withing `examples/server/public`.

Is it possible to fix or undo this?

mounta11n committed:
add tinyllama chat prompt

AmgadHasan commented:
Hey @mounta11n 
Thanks for working on this. The new UI looks really good. Any updates about the PR?

ggerganov commented:
A new UI has already been merged via https://github.com/ggerganov/llama.cpp/pull/7633
